Feature 채팅방 별 익명의 닉네임 반환하기

slowCoyle opened the pull request:
## Resolve #188 

- 채팅방 별 중복되지 않는 익명의 닉네임이 필요하다.
- 연결 종료 시 부여받은 닉네임을 제거하는 기능이 필요하다.

## Changes

- 채팅방 별 중복되지 않는 익명의 닉네임 반환
- 채팅방 connection 종료 시 닉네임과 noticeId로 해당하는 닉네임 제거

## Notes

- 동사 + 명사의 형태로 익명의 닉네임을 부여한다.

## References

- [Value Collection 저장](https://javacan.tistory.com/tag/%40CollectionTable)

Commit:
feat: 채팅방이 존재하지 않은 경우 채팅방 생성하는 PATCH API 작성

Commit:
feat: 채팅방 소켓 connect 시 채팅방 생성하기 API 연결

Commit:
feat: 중복되지 않은 익명 채팅 이름 받아오기

Commit:
feat: 채팅방 Connection 시 닉네임을 포함한 채팅방 정보 가져오기

- 닉네임은 중복되지 않도록 구현하였다.

Commit:
feat: disConnection 시 공고 ID로 닉네임 삭제 서비스 기능 구현

Commit:
test: 도메인 방어 테스트 작성

Commit:
feat:Notice Id, Nickname으로 NickName 삭제 Controller 기능

KimSeongGyu1 commented on build.gradle:
```diff
@@ -91,7 +91,7 @@ dependencies {
     compile group: 'com.googlecode.json-simple', name: 'json-simple', version: '1.1'
     compile group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.5.12'
-    developmentOnly 'org.springframework.boot:spring-boot-devtools'
+    // developmentOnly 'org.springframework.boot:spring-boot-devtools'
```
이 부분은 왜 주석처리했나요?

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/controller/ChatController.java:
```diff
@@ -24,8 +28,16 @@ public void sendMessage(MessageSendRequest messageSendRequest) {
-    @GetMapping("/api/chats")
-    public ResponseEntity<ChatRoomResponse> fetchChatRoom(@RequestParam(name = "noticeId") Long noticeId) {
-        return ResponseEntity.ok().body(chatService.fetchChatRoom(noticeId));
+    @PatchMapping("/api/chatrooms")
+    public ResponseEntity<ChatRoomResponse> createIfNotExist(@RequestBody ChatRoomCreateRequest chatRoomCreateRequest) {
+        return ResponseEntity.ok().body(chatService.createIfNotExist(chatRoomCreateRequest));
+    }
```
url이 /api/chatrooms/{id} 까지 있는 것이 patch에 대해 더 rest스럽다고 생각하는데 어떻게 생각하세요?

KimSeongGyu1 replied:
/api/chatrooms/{noticeId}도 괜찮지 않을까라고 생각했는데
다시 생각해보니깐 굉장히 별로네요

차라리 /api/chatrooms?noticeId= 는 어때요?

slowCoyle replied:
오 좋은 생각 같네요! 검색이 기반이 되는 url 요청이니까요. 수정했습니다.

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/Verb.java:
```diff
@@ -0,0 +1,6 @@
+public enum Verb {
+
+    만지는, 돌리는, 마시는
+}
```
핸드폰, 커피 등의 명사와 함께 사용한다면
Verb보다는 Adjective를 활용하는 것이 더 자연스러울 것 같아요

구글에 형용사 모음이라고 검색하니 이런 것들이 엄청 많아요 참고하면 좋을 것 같아요
![image](https://user-images.githubusercontent.com/56247080/90967587-2a50ba00-e51c-11ea-9fe8-38c4db8b4cdc.png)

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/service/ChatService.java:
```diff
@@ -26,27 +29,41 @@
+    }
 
-        return ChatRoomResponse.of(chatRoom.getChats(), "홍길동");
+    public void deleteNickName(String nickName, Long noticeId) {
```
@Transactional을 걸고 save()를 호출하지 않는 것이 JPA를 좀 더 잘 활용하는 방향일 것 같아요

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/AllChatNames.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Collections;
+import java.util.List;
+
+public class AllChatNames {
```
이름이 ChatNameFactory였으면 더 보기 편했을 것 같아요

KimSeongGyu1 replied:
nonOverlappingChatNameFactory는 좀 길어보여요ㅜ
개인적으로는 ChatNameFacatory가 낫지만 코일이 하고 싶으면 해도 된다고 생각해요

slowCoyle replied:
ChatNameFactory로 수정했습니다!

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/AllChatNames.java:
```diff
@@ -0,0 +1,34 @@
+
+    private static final List<ChatName> names = createAllChatNames();
+
+    public static List<ChatName> createAllChatNames() {
```
외부에서 사용할 일이 없을 것 같은데 private인 편이 좋지 않을까요?

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/AllChatNames.java:
```diff
@@ -0,0 +1,34 @@
+
+public class AllChatNames {
+
+    private static final List<ChatName> names = createAllChatNames();
```
닉네임의 순서는 중요하지 않으니 Set을 사용하면 중복이 없을 것이라는 것을 더 명확히 표현할 수 있을 것 같아요
혹시 모를 중복의 발생도 예방할 수 있고요

괜찮다면 다른 곳까지 전반적으로 수정할 수 있을까요?

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/AllChatNames.java:
```diff
@@ -0,0 +1,34 @@
+        return chatNames;
+    }
+
+    public static ChatName fetchNonRedundantName(List<ChatName> existNameInChatRoom) {
```
fetch보다는 create, generate라는 단어가 더 잘 어울릴 것 같아요

`redendunt`는 `중복적이다`라는 뜻도 있지만 `불필요하다`라는 느낌이 강하다고 느껴져요
nonRedendunt보다는 nonOverlapping이 더 좋을 것 같은데 어때요?

KimSeongGyu1 commented on src/frontend/src/store/modules/chat.js:
```diff
@@ -15,6 +15,7 @@ export default {
       const socket = new SockJS("/chat");
       state.stompClient = Stomp.over(socket);
       state.stompClient.connect({}, function(frame) {
+        patchAction("/api/chatrooms", { noticeId: noticeId }, {});
```
이 부분은 직접 테스트하시느라 넣으신 것 같은데
제가 프론트에서 관련 작업을 하는 것이 좀 더 작업이 원활할 것 같아요
pr에서는 빼도 될까요?

KimSeongGyu1 replied:
넵

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/ChatName.java:
```diff
@@ -0,0 +1,27 @@
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+@EqualsAndHashCode
+@Embeddable
```
Embeddable이 제일 위로 가는 것이 팀 컨벤션에 맞는 것 같아요

KimSeongGyu1 commented on src/main/java/underdogs/devbie/chat/domain/ChatNames.java:
```diff
@@ -0,0 +1,45 @@
+@Getter
+public class ChatNames {
+
+    @ElementCollection(fetch = FetchType.EAGER)
```
EAGER하게 컬렉션을 들고 온다면 n+1문제가 걱정되는데
혹시 JPA에서 생성되는 쿼리를 확인해주실수 있나요

KimSeongGyu1 replied:
코일이 말한대로 지금 chatRoom은 chatNames로 무조건 이어지므로, fetch join을 활용해서 간단한 CRUD는 직접 JPQL을 작성하는 것이 괜찮다고 생각해요.

이때 엔티티를 사용한다면 코일이 말한대로 하면 U의 문제를 어느정도 해결할 수는 있지만 C,R,D에서 발생하는 n+1 문제는 해결하지 못한다고 알고 있어요.
그래서 엔티티를 사용하는 것은 괜찮지만 사용하더라도 fetch join은 활용해야 한다고 생각해요

KimSeongGyu1 commented on src/test/java/underdogs/devbie/MvcTest.java:
```diff
@@ -78,6 +78,14 @@ protected ResultActions putAction(String url, String inputJson, String bearerTok
                 .contentType(MediaType.APPLICATION_JSON));
     }
 
+    protected ResultActions patchAction(String url, String inputJson) throws Exception {
+        return this.mockMvc
+            .perform(patch(url)
+                .content(inputJson)
+                .accept(MediaType.APPLICATION_JSON)
+                .contentType(MediaType.APPLICATION_JSON));
+    }
+
```
기존의 메서드에 bearerToken을 빈 String으로 넣어도 원하는 동작이 가능할텐데
따로 메서드를 빼는 것이 더 좋을까요?

KimSeongGyu1 replied:
그렇네요 
코일이 한 방향이 더 휴먼 에러를 줄이는 방향이 될 것 같아요
배워갑니다

KimSeongGyu1 reviewed (commented):
고생하셨어요

코멘트 남긴 것들은 한번 같이 생각해보고 싶어서 남긴 것들이니
무조건 반영하는 것 보다 의견 남겨주면 좋을 것 같아요

그리고 파일 몇개에 마지막 줄은 한 줄 띄어주면 좋을 것 같아요

Commit:
refactor: dirty checking을 위한 @Transactional 제거

Commit:
refactor: ChatNames 자료구조 Set으로 변경

- 순서의 중요성이 없다고 판단해 Set으로 변경

Commit:
refactor: 네이밍 변경

- Verb(동사)에서 Adjective(형용사)로 변경
- AllChatNames에서 ChatNameFactory로 변경

Commit:
refactor: build.gradle devtools 주석 제거

Commit:
refactor: vue 클라이언트 채팅방 생성 patch api 호출 메서드 제거

Commit:
refactor: ChatRoom Patch URL 변경

- body에 담긴 noticeId를 통해 patch할 chatroom을 식별하였는데
식별에 의미를 둔다면 RequestParam 을 통해 식별하는 것이 맞는
URL이라고 판단하였다.

Commit:
refactor: ChatName, ChatMessage Embedded 타입 사용

Commit:
feat: ChatName 별 TitleColor 생성 Controller Test 작성

Commit:
feat: ChatName에 TitleColor 를 가지도록 변경

Commit:
feat: TitleColor, TitleColorFactory 추가

Commit:
feat: 채팅방 인원수 반환 ChatControllerTest 추가

Commit:
feat: 채팅방 인원수 반환 ChatServiceTest 추가 및 구현

KimSeongGyu1 reviewed (approved):
고생했어요
파일 맨 뒤에 빈 라인 추가 안된 부분만 추가하고 머지해주세요

n+1문제는 기록해두었다가 다음에 해결해주세요

Commit:
refactor: no newline at end of file 해결
